### ISSUE TITLE
Open to docker docs improvements?

### ISSUE THREAD

**Issue by chmac** [jelmer/xandikos#144]
I'm looking to run xandikos in docker. I'm not sure from the readme which paths I need to supply to docker, what ports it exposes, how SSL termination works, etc, etc. Assuming I can figure this stuff out, are you open to a PR to expand on the docker docs? I might also include an example docker-compose file.

**Comment by chmac** [comment]
I'll start gathering some notes here which I can turn into a PR later if you're in favour.

Example compose file

```yaml
version: "3.4"

services:
  xandikos:
    image: ghcr.io/jelmer/xandikos
    ports:
      - 8000:8000
    volumes:
      - /path/to/xandikos/data:/data
    restart: unless-stopped
```

Notes

- If you want to add data from an external tool, add it to git and commit the changes then xandikos will pick up the changes
- The docker container needs some kind of setup setup to create the address book, etc
  - `docker exec -it CONTAINER_NAME /code/bin/xandikos --defaults -d /data` probably works
- The docker container listens on port 8000 with no authentication
  - Point a reverse proxy here + add any authentication in front
- Quick & dirty way to test, use Thunderbird & the CardBook plugin (plus ssh port forwards if necessary)
- Filenames in the addressbook directory are the contact's UUID plus the `.vcf` extension

**Comment by jelmer** [comment]
Yeah, improvements to the documentation would be very welcome. I'm not sure about docker-compose though - I don't use it myself; so I'm open to merging it, but a little worried about keeping it up to date.

**Comment by chmac** [comment]
@jelmer Awesome.

The docker compose just says "run this container with these directories passed in", so it shouldn't require any maintenance. The main thing is it tells people which ports and which directories they should pay attention to.

**Comment by jelmer** [comment]
You shouldn't need a setup to creat the addressbook, etc. It's sufficient to just specify --defaults (which is done by default).

You can directly modify the git repository, but I wouldn't want to recommend that as a common way to import changes - it's much better to just use the CalDAV interface.

The filenames for e.g. vcf files depend on what the clients do - there isn't a specific format that xandikos mandates.

**Comment by chmac** [comment]
@jelmer Okay, thanks for the clarification. I'm working on a markdown -> vcf sync tool, which one day I might want to work bidirectionally. But for now I'm thinking to just push data into xandikos and then use it to serve my mobile phone, as the server.

Are the default folders created when xandikos first runs? Or does it need to receive a request first? I found that I didn't have anything in my `/data` folder until I ran the command  manually inside the container.

**Comment by jelmer** [comment]
You can directly commit vcf files to the xandikos git repository but  that means you don't get any validation for the files, i.e. xandikos doesn't get a chance to reject files that are invalid for some reason.

An advantage of using CalDAV is also that you can use it with other CalDAV servers that e.g. store the files in a database of some sort.

The default folders are created if you specify the --defaults argument.

**Comment by chmac** [comment]
@jelmer True, you make a compelling argument with the idea to use the server via the CalDAV protocol. I could then use any server I want. Hmm, will think about that. I figured files & git would be faster, and I know them, but you're definitely right about the portability. Thanks for the suggestion.

**Comment by jelmer** [comment]
So you could check files directly into git and it would indeed be faster, but I'd rather not promote that as the primary way of getitng data into Xandikos. Unless you're talking hundreds of files, it's probably not going to be that much slower either to use the CalDAV interface.

**Comment by chmac** [comment]
@jelmer Makes sense. That note was mostly for me, I can totally understand it doesn't really make sense to push to git directly. I ended up implementing a CardDAV client as you suggested, which has worked out great, as now I can push my contacts into any server, including the one used by my email provider!

I'm happy to close this issue now. #148 was merged, and I think between that and the new man page, it covers all of the topics that came up for me.

Thanks for being so responsive to the input, I really appreciate it.
